File: abxbus-core/crates/abxbus/src/event_history.rs
use std::sync::Arc;

use indexmap::IndexMap;
use tracing::warn;

use crate::event::BaseEvent;
use crate::monotonic_dt::monotonic_datetime;
use crate::types::EventStatus;

/// Ordered event history map with query and trim helpers.
///
/// Events are stored in insertion order (FIFO) using an IndexMap.
/// Equivalent to Python's `EventHistory(dict[str, BaseEvent])`.
pub struct EventHistory {
    /// Event storage: event_id -> event.
    events: IndexMap<String, Arc<BaseEvent>>,
    /// Maximum number of events to keep in history.
    /// None = unlimited, 0 = keep only in-flight events.
    pub max_history_size: Option<usize>,
    /// If true, drop oldest events when full. If false, reject new events.
    pub max_history_drop: bool,
    /// Whether we've warned about dropping uncompleted events.
    warned_about_dropping: bool,
}

impl EventHistory {
    pub fn new(max_history_size: Option<usize>, max_history_drop: bool) -> Self {
        Self {
            events: IndexMap::new(),
            max_history_size,
            max_history_drop,
            warned_about_dropping: false,
        }
    }

    pub fn len(&self) -> usize {
        self.events.len()
    }

    pub fn is_empty(&self) -> bool {
        self.events.is_empty()
    }

    pub fn add_event(&mut self, event: Arc<BaseEvent>) {
        self.events.insert(event.event_id.clone(), event);
    }

    pub fn get_event(&self, event_id: &str) -> Option<&Arc<BaseEvent>> {
        self.events.get(event_id)
    }

    pub fn remove_event(&mut self, event_id: &str) -> bool {
        self.events.shift_remove(event_id).is_some()
    }

    pub fn has_event(&self, event_id: &str) -> bool {
        self.events.contains_key(event_id)
    }

    pub fn clear(&mut self) {
        self.events.clear();
    }

    pub fn values(&self) -> impl Iterator<Item = &Arc<BaseEvent>> {
        self.events.values()
    }

    pub fn iter(&self) -> impl Iterator<Item = (&String, &Arc<BaseEvent>)> {
        self.events.iter()
    }

    /// Normalize an event pattern to a string key.
    pub fn normalize_event_pattern(event_pattern: &str) -> &str {
        if event_pattern == "*" {
            "*"
        } else {
            event_pattern
        }
    }

    /// Check if `event` is a descendant of `ancestor` by walking the parent chain.
    pub fn event_is_child_of(&self, event: &BaseEvent, ancestor: &BaseEvent) -> bool {
        let mut current_id = event.event_parent_id.as_deref();
        let mut visited = std::collections::HashSet::new();

        while let Some(cid) = current_id {
            if visited.contains(cid) {
                break;
            }
            if cid == ancestor.event_id {
                return true;
            }
            visited.insert(cid.to_string());
            current_id = self
                .events
                .get(cid)
                .and_then(|e| e.event_parent_id.as_deref());
        }

        false
    }

    /// Find the most recent event matching criteria in history.
    ///
    /// - `event_key`: event type string or "*" for all
    /// - `past`: true/false or seconds lookback window
    /// - `where_fn`: optional predicate filter
    /// - `child_of`: optional ancestor constraint
    /// - `field_filters`: exact-match filters on extra_fields
    pub fn find_in_past(
        &self,
        event_key: &str,
        past_seconds: Option<f64>,
        where_fn: Option<&dyn Fn(&BaseEvent) -> bool>,
        child_of: Option<&BaseEvent>,
        field_filters: &[(&str, &serde_json::Value)],
    ) -> Option<Arc<BaseEvent>> {
        let cutoff = past_seconds.map(|secs| {
            let cutoff_dt = chrono::Utc::now() - chrono::Duration::milliseconds((secs * 1000.0) as i64);
            monotonic_datetime(Some(&cutoff_dt.format("%Y-%m-%dT%H:%M:%S%.9fZ").to_string()))
                .unwrap_or_default()
        });

        // Iterate in reverse (most recent first).
        for event in self.events.values().rev() {
            // Cutoff check
            if let Some(ref cutoff_str) = cutoff {
                if event.event_created_at < *cutoff_str {
                    continue;
                }
            }

            // Pattern match
            if event_key != "*" && event.event_type != event_key {
                continue;
            }

            // Child-of check
            if let Some(ancestor) = child_of {
                if !self.event_is_child_of(event, ancestor) {
                    continue;
                }
            }

            // Field filters
            let mut fields_match = true;
            for (field_name, expected_value) in field_filters {
                let actual = event.extra_fields.get(*field_name);
                if actual != Some(expected_value) {
                    // Also check core fields
                    let core_match = match *field_name {
                        "event_type" => serde_json::Value::String(event.event_type.clone()) == **expected_value,
                        "event_id" => serde_json::Value::String(event.event_id.clone()) == **expected_value,
                        "event_status" => serde_json::Value::String(event.event_status.to_string()) == **expected_value,
                        _ => false,
                    };
                    if !core_match {
                        fields_match = false;
                        break;
                    }
                }
            }
            if !fields_match {
                continue;
            }

            // Custom predicate
            if let Some(pred) = where_fn {
                if !pred(event) {
                    continue;
                }
            }

            return Some(event.clone());
        }

        None
    }

    /// Trim event history to stay within max_history_size.
    /// Returns the number of events removed.
    pub fn trim_event_history(&mut self, owner_label: Option<&str>) -> usize {
        let max_size = match self.max_history_size {
            None => return 0,
            Some(0) => {
                // max_history_size=0: remove all completed events.
                let completed_ids: Vec<String> = self
                    .events
                    .iter()
                    .filter(|(_, e)| e.event_status == EventStatus::Completed)
                    .map(|(id, _)| id.clone())
                    .collect();
                let count = completed_ids.len();
                for id in completed_ids {
                    self.events.shift_remove(&id);
                }
                return count;
            }
            Some(size) => size,
        };

        if !self.max_history_drop || self.events.len() <= max_size {
            return 0;
        }

        let mut remaining_overage = self.events.len() - max_size;
        let mut removed_count = 0;

        // First pass: remove completed events.
        let completed_ids: Vec<String> = self
            .events
            .iter()
            .filter(|(_, e)| e.event_status == EventStatus::Completed)
            .take(remaining_overage)
            .map(|(id, _)| id.clone())
            .collect();

        for id in completed_ids {
            self.events.shift_remove(&id);
            removed_count += 1;
            remaining_overage -= 1;
            if remaining_overage == 0 {
                return removed_count;
            }
        }

        // Second pass: remove any events (including uncompleted).
        if remaining_overage > 0 {
            let all_ids: Vec<String> = self
                .events
                .keys()
                .take(remaining_overage)
                .cloned()
                .collect();
            let dropped_uncompleted = all_ids
                .iter()
                .filter(|id| {
                    self.events
                        .get(*id)
                        .is_some_and(|e| e.event_status != EventStatus::Completed)
                })
                .count();

            for id in all_ids {
                self.events.shift_remove(&id);
                removed_count += 1;
            }

            if dropped_uncompleted > 0 && !self.warned_about_dropping {
                self.warned_about_dropping = true;
                let owner = owner_label.unwrap_or("EventBus");
                warn!(
                    "[abxbus] ⚠️ Bus {} has exceeded max_history_size={} and is dropping oldest history entries \
                     (even uncompleted events). Increase max_history_size or set max_history_drop=false to reject.",
                    owner, max_size
                );
            }
        }

        removed_count
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_event(event_type: &str) -> Arc<BaseEvent> {
        Arc::new(BaseEvent::new(event_type))
    }

    #[test]
    fn test_add_and_get() {
        let mut history = EventHistory::new(Some(100), false);
        let event = make_event("TestEvent");
        let id = event.event_id.clone();
        history.add_event(event.clone());
        assert!(history.has_event(&id));
        assert_eq!(history.get_event(&id).unwrap().event_type, "TestEvent");
    }

    #[test]
    fn test_find_in_past() {
        let mut history = EventHistory::new(Some(100), false);
        let e1 = make_event("TypeA");
        let e2 = make_event("TypeB");
        let e3 = make_event("TypeA");
        history.add_event(e1);
        history.add_event(e2);
        history.add_event(e3.clone());

        let found = history.find_in_past("TypeA", None, None, None, &[]);
        assert!(found.is_some());
        // Should find the most recent TypeA (e3).
        assert_eq!(found.unwrap().event_id, e3.event_id);

        let found_b = history.find_in_past("TypeB", None, None, None, &[]);
        assert!(found_b.is_some());

        let found_none = history.find_in_past("TypeC", None, None, None, &[]);
        assert!(found_none.is_none());

        // Wildcard
        let found_all = history.find_in_past("*", None, None, None, &[]);
        assert!(found_all.is_some());
    }

    #[test]
    fn test_trim_completed() {
        let mut history = EventHistory::new(Some(2), true);
        for i in 0..5 {
            let mut event = BaseEvent::new(format!("Event{}", i));
            event.event_status = EventStatus::Completed;
            event.is_complete = true;
            history.add_event(Arc::new(event));
        }
        assert_eq!(history.len(), 5);
        let removed = history.trim_event_history(Some("TestBus"));
        assert!(removed > 0);
        assert!(history.len() <= 2);
    }
}

File: abxbus-ts/src/EventHistory.ts
import { BaseEvent } from './BaseEvent.js'
import type { EventPattern, FindWindow } from './types.js'
import { normalizeEventPattern } from './types.js'
import { monotonicDatetime } from './helpers.js'

const normalizeMaxHistorySize = (value: number | null | undefined): number | null => {
  if (value === undefined) {
    return 100
  }
  if (value === null || value <= 0) {
    return null
  }
  return value
}

export type EventHistoryFindOptions = {
  past?: FindWindow
  future?: FindWindow
  child_of?: BaseEvent | null
  event_is_child_of?: (event: BaseEvent, ancestor: BaseEvent) => boolean
  wait_for_future_match?: (
    event_pattern: string | '*',
    matches: (event: BaseEvent) => boolean,
    future: FindWindow
  ) => Promise<BaseEvent | null>
} & Record<string, unknown>

export type EventHistoryFilterOptions = EventHistoryFindOptions & { limit?: number | null }

export type EventHistoryTrimOptions<TEvent extends BaseEvent = BaseEvent> = {
  is_event_complete?: (event: TEvent) => boolean
  on_remove?: (event: TEvent) => void
  owner_label?: string
  max_history_size?: number | null
  max_history_drop?: boolean
}

export class EventHistory<TEvent extends BaseEvent = BaseEvent> implements Iterable<[string, TEvent]> {
  max_history_size: number | null
  max_history_drop: boolean

  private _events: Map<string, TEvent>
  private _warned_about_dropping_uncompleted_events: boolean

  constructor(options: { max_history_size?: number | null; max_history_drop?: boolean } = {}) {
    this.max_history_size = normalizeMaxHistorySize(options.max_history_size)
    this.max_history_drop = options.max_history_drop ?? false
    this._events = new Map()
    this._warned_about_dropping_uncompleted_events = false
  }

  get size(): number {
    return this._events.size
  }

  [Symbol.iterator](): Iterator<[string, TEvent]> {
    return this._events[Symbol.iterator]()
  }

  entries(): IterableIterator<[string, TEvent]> {
    return this._events.entries()
  }

  keys(): IterableIterator<string> {
    return this._events.keys()
  }

  values(): IterableIterator<TEvent> {
    return this._events.values()
  }

  clear(): void {
    this._events.clear()
  }

  get(event_id: string): TEvent | undefined {
    return this._events.get(event_id)
  }

  set(event_id: string, event: TEvent): this {
    this._events.set(event_id, event)
    return this
  }

  has(event_id: string): boolean {
    return this._events.has(event_id)
  }

  delete(event_id: string): boolean {
    return this._events.delete(event_id)
  }

  addEvent(event: TEvent): void {
    this._events.set(event.event_id, event)
  }

  getEvent(event_id: string): TEvent | undefined {
    return this._events.get(event_id)
  }

  removeEvent(event_id: string): boolean {
    return this._events.delete(event_id)
  }

  hasEvent(event_id: string): boolean {
    return this._events.has(event_id)
  }

  static normalizeEventPattern(event_pattern: EventPattern | '*'): string | '*' {
    return normalizeEventPattern(event_pattern)
  }

  find(event_pattern: '*', where?: (event: TEvent) => boolean, options?: EventHistoryFindOptions): Promise<TEvent | null>
  find<TMatch extends TEvent>(
    event_pattern: EventPattern<TMatch>,
    where?: (event: TMatch) => boolean,
    options?: EventHistoryFindOptions
  ): Promise<TMatch | null>
  async find(
    event_pattern: EventPattern<TEvent> | '*',
    where: (event: TEvent) => boolean = () => true,
    options: EventHistoryFindOptions = {}
  ): Promise<TEvent | null> {
    // `limit` field-equality filter would collide with filter()'s cap arg; route it through `where`.
    let effective_where = where
    let effective_options: EventHistoryFindOptions = options
    if (Object.prototype.hasOwnProperty.call(options, 'limit')) {
      const { limit: limit_field_value, ...rest } = options as EventHistoryFindOptions & { limit: unknown }
      const inner_where = where
      effective_where = (event: TEvent) => (event as unknown as Record<string, unknown>).limit === limit_field_value && inner_where(event)
      effective_options = rest as EventHistoryFindOptions
    }
    const results = await this.filter(event_pattern as EventPattern<TEvent> | '*', effective_where, {
      ...effective_options,
      limit: 1,
    })
    return results.length > 0 ? results[0] : null
  }

  filter(event_pattern: '*', where?: (event: TEvent) => boolean, options?: EventHistoryFilterOptions): Promise<TEvent[]>
  filter<TMatch extends TEvent>(
    event_pattern: EventPattern<TMatch>,
    where?: (event: TMatch) => boolean,
    options?: EventHistoryFilterOptions
  ): Promise<TMatch[]>
  async filter(
    event_pattern: EventPattern<TEvent> | '*',
    where: (event: TEvent) => boolean = () => true,
    options: EventHistoryFilterOptions = {}
  ): Promise<TEvent[]> {
    const past = options.past ?? true
    const future = options.future ?? false
    const child_of = options.child_of ?? null
    const eventIsChildOf = options.event_is_child_of ?? ((event: BaseEvent, ancestor: BaseEvent) => this.eventIsChildOf(event, ancestor))
    const waitForFutureMatch = options.wait_for_future_match
    const limit = options.limit ?? null
    if (past === false && future === false) {
      return []
    }

    if (limit !== null && limit <= 0) {
      return []
    }

    const event_key = EventHistory.normalizeEventPattern(event_pattern)
    const cutoff_at = past === true ? null : monotonicDatetime(new Date(Date.now() - Math.max(0, Number(past)) * 1000).toISOString())

    const event_field_filters = Object.entries(options).filter(
      ([key, value]) =>
        key !== 'past' &&
        key !== 'future' &&
        key !== 'child_of' &&
        key !== 'event_is_child_of' &&
        key !== 'wait_for_future_match' &&
        key !== 'limit' &&
        value !== undefined
    )

    const matches = (event: BaseEvent): boolean =>
      (event_key === '*' || event.event_type === event_key) &&
      (!child_of || eventIsChildOf(event, child_of)) &&
      event_field_filters.every(([field_name, expected]) => (event as unknown as Record<string, unknown>)[field_name] === expected) &&
      where(event as TEvent)

    const results: TEvent[] = []
    if (past !== false) {
      const history_values = Array.from(this._events.values())
      for (let i = history_values.length - 1; i >= 0; i -= 1) {
        const event = history_values[i]
        if (cutoff_at !== null && event.event_created_at < cutoff_at) {
          continue
        }
        if (matches(event)) {
          results.push(event)
          if (limit !== null && results.length >= limit) {
            return results
          }
        }
      }
    }

    if (future === false || !waitForFutureMatch) {
      return results
    }

    const future_match = (await waitForFutureMatch(event_key, matches, future)) as TEvent | null
    if (future_match !== null) {
      results.push(future_match)
    }
    return results
  }

  trimEventHistory(options: EventHistoryTrimOptions<TEvent> = {}): number {
    const max_history_size = normalizeMaxHistorySize(options.max_history_size ?? this.max_history_size)
    const max_history_drop = options.max_history_drop ?? this.max_history_drop
    if (max_history_size === null) {
      return 0
    }

    const is_event_complete = options.is_event_complete ?? ((event: TEvent) => event.event_status === 'completed')
    const on_remove = options.on_remove

    if (!max_history_drop || this.size <= max_history_size) {
      return 0
    }

    let remaining_overage = this.size - max_history_size
    let removed_count = 0
    const remove_event = (event_id: string, event: TEvent): void => {
      this._events.delete(event_id)
      on_remove?.(event)
      removed_count += 1
    }

    for (const [event_id, event] of Array.from(this._events.entries())) {
      if (remaining_overage <= 0) {
        break
      }
      if (!is_event_complete(event)) {
        continue
      }
      remove_event(event_id, event)
      remaining_overage -= 1
    }

    let dropped_uncompleted = 0
    for (const [event_id, event] of Array.from(this._events.entries())) {
      if (remaining_overage <= 0) {
        break
      }
      if (!is_event_complete(event)) {
        dropped_uncompleted += 1
      }
      remove_event(event_id, event)
      remaining_overage -= 1
    }

    if (dropped_uncompleted > 0 && !this._warned_about_dropping_uncompleted_events) {
      this._warned_about_dropping_uncompleted_events = true
      const owner_label = options.owner_label ?? 'EventBus'
      console.error(
        `[abxbus] ⚠️ Bus ${owner_label} has exceeded max_history_size=${max_history_size} and is dropping oldest history entries (even uncompleted events). Increase max_history_size or set max_history_drop=false to reject.`
      )
    }

    return removed_count
  }

  private eventIsChildOf(event: BaseEvent, ancestor: BaseEvent): boolean {
    let current_parent_id = event.event_parent_id
    const visited = new Set<string>()

    while (current_parent_id && !visited.has(current_parent_id)) {
      if (current_parent_id === ancestor.event_id) {
        return true
      }
      visited.add(current_parent_id)
      const parent = this._events.get(current_parent_id)
      if (!parent) {
        return false
      }
      current_parent_id = parent.event_parent_id
    }

    return false
  }
}
